multicluster/controllers/multicluster/core/leader_elector.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2021 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"crypto/rand"
+	"math/big"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+/**
+ * Lets consider a ClusterSet with 2 leader clusters and 1 member cluster
+ *      Leader1        Leader2
+ *	     \         /
+ *	      \       /
+ *             -------
+ *               ^
+ *               |
+ *            Member1
+ * The member cluster connects to both leaders and performs Member Announce
+ * to announce itself. And all resources from the Members are written into
+ * Common Area of all Leader clusters. However resources are imported
+ * only from an elected-leader cluster.
+ *
+ * This interface performs the election to pick an elected-leader
+ * from the list of leader clusters in the ClusterSet spec.
+ *
+ * Leader Election picks an elected-leader among "connected" leaders
+ * by choosing one randomly
+ * The Member Cluster periodically writes MemberAnnounce CRD into every leader
+ * cluster and is considered "connected" if it can successfully write so.
+ * After leader election is done, the result is also used to update the
+ * spec of the MemberAnnounce CRD during the next periodic write into
+ * the leader's Common Area, so the leader knows it is the elected-leader
+ * of this member (mainly for visibility and reporting)
+ */
+type LeaderElector interface {
+	// start the leader elector, returns a cancelFunc to invoke when it must be stopped
+	StartLeaderElection() context.CancelFunc
+}
+
+func (m *remoteClusterManager) StartLeaderElection() (context.CancelFunc, error) {
+	stopCtx, stopFunc := context.WithCancel(context.Background())
+
+	// Start a Timer for every 5 seconds when leader election is
+	// performed, if necessary
+	ticker := time.NewTicker(5 * time.Second)
+
+	go func() {
+		klog.Infof("Starting leader election")
+		for {
+			select {
+			case <-stopCtx.Done():
+				klog.Infof("Stopping leader election")
+				return
+			case <-ticker.C:
+				// 5 second timer has gone off, check if election must be done
+				if !m.needElection {
```

Comment:
Do we need locking for accessing remoteClusterManager and remoteCluster attributes?

Reply:
Yes because they are running in 2 different goroutine. Let me consolidate the goroutine in LeaderElector and send another patch and I can see if I can eliminate any locks.

Reply:
Pushed a patch to do this now, please let me know if it looks better.